src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -1191,21 +1138,59 @@ public ActivityIdFormat IdFormat
             private set => _state = (_state & ~State.FormatFlags) | (State)((byte)value & (byte)State.FormatFlags);
         }
 
-        /// <summary>
-        /// Having our own key-value linked list allows us to be more efficient
-        /// </summary>
         private partial class LinkedListNode<T>
         {
             public LinkedListNode(T value) => Value = value;
-            public LinkedListNode(T value, LinkedListNode<T>? next)
-            {
-                Value = value;
-                Next = next;
-            }
             public T Value;
             public LinkedListNode<T>? Next;
         }
 
+        private class LinkedList<T>
+        {
+            private LinkedListNode<T> _first;
+            private LinkedListNode<T> _last;
+
+            public LinkedList(T firstValue) => _last =_first = new LinkedListNode<T>(firstValue);
+
+            public LinkedList(IEnumerable<T> list)
+            {
+                IEnumerator<T> enumerator = list.GetEnumerator();
+                Debug.Assert(enumerator.MoveNext());
+                _last =_first = new LinkedListNode<T>(enumerator.Current);
+
+                while (enumerator.MoveNext())
+                {
+                    _last.Next = new LinkedListNode<T>(enumerator.Current);
+                    _last = _last.Next;
+                }
+            }
+
+            public LinkedListNode<T> First => _first;
+
+            public void Add(T value)
+            {
+                LinkedListNode<T> newNode = new LinkedListNode<T>(value);
+                LinkedListNode<T> last;
+
+                do
+                {
+                    last = _last;
+                } while (!object.ReferenceEquals(Interlocked.CompareExchange(ref _last, newNode, last), last));
+
+                last.Next = newNode;
```

Comment:
It looks like this will eventually be consistent, but it may not yield expected results.

- Let's say the list currently has two nodes, A -> B, so _first is A and _last is B.
- Thread 1 Adds C, setting _last to C but getting context switched out before it gets a chance to set last.Next = B.
- Thread 2 Adds D, setting _last to D and last.Next = B; here last will be C.
- Thread 2 calls Enumerate.  It'll see A and B, but it won't see C or D, even though it just successfully added D.

Reply:
I had seen this earlier and convinced myself the eventual consistency would be fine because synchronization outside the API but required to use it properly would force consistency, preventing a reader from ever observing it in the inconsistent state. Now your comment made me think about it a lot more and reverse my opinion : )

I suspect this could be remedied doing the CompareExchange on the Next pointer of the final node, then do an unsynchronized write to update _last if the swap was successful. To find the final node a thread would need to start from _last and then iterate until Next == null. It is possible that _last will not point to the final node of the linked list, and it may not become eventually consistent, but that wouldn't interfere with the functional correctness of the API. At worst it creates a slight perf penalty for the next Add() operation.